src/javascript/waxeye.ts

```diff
@@ -0,0 +1,734 @@
+
+export interface EmptyAST extends AST {
+  type: '';
+  children: never[];
+}
+
+export function EmptyAST(): EmptyAST {
+  return new AST('', []) as EmptyAST;
+}
+
+export type Expr = ExprNonTerminal|ExprAlt|ExprSeq|ExprPlus|ExprStar|ExprOpt|
+    ExprAnd|ExprNot|ExprVoid|ExprAny|ExprChar|ExprCharClass;
+
+export const enum ExprType {
+  NT = 1,
+  ALT,
+  SEQ,
+  PLUS,
+  STAR,
+  OPT,
+  AND,
+  NOT,
+  VOID,
+  ANY,
+  CHAR,
+  CHAR_CLASS,
+}
+
+export interface ExprNonTerminal {
+  type: ExprType.NT;
+  // Non-terminal's name.
+  args: [string];
+}
+export interface ExprAlt {
+  type: ExprType.ALT;
+  args: Expr[];
+}
+export interface ExprSeq {
+  type: ExprType.SEQ;
+  args: Expr[];
+}
+export interface ExprPlus {
+  type: ExprType.PLUS;
+  args: [Expr];
+}
+export interface ExprStar {
+  type: ExprType.STAR;
+  args: [Expr];
+}
+export interface ExprOpt {
+  type: ExprType.OPT;
+  args: [Expr];
+}
+export interface ExprAnd {
+  type: ExprType.AND;
+  args: [Expr];
+}
+export interface ExprNot {
+  type: ExprType.NOT;
+  args: [Expr];
+}
+export interface ExprVoid {
+  type: ExprType.VOID;
+  args: [Expr];
+}
+export interface ExprAny {
+  type: ExprType.ANY;
+  args: never[];
+}
+export interface ExprChar {
+  type: ExprType.CHAR;
+  // A single character (represented by a single Unicode codepoint).
+  args: [string];
+}
+export interface ExprCharClass {
+  type: ExprType.CHAR_CLASS;
+  // A list of Unicode codepoints / ranges of codepoints.
+  args: Array<number|[number, number]>;
+}
+
+export interface MatchError {
+  // Expected expression in the grammar format.
+  toGrammarString(): string;
+}
+
+// A failed single character match.
+export class ErrChar implements MatchError {
+  constructor(public char: string) {}
+
+  public toGrammarString() {
+    return `'${JSON.stringify(this.char).slice(1, -1)}'`;
+  }
+}
+
+// A failed character class match.
+export class ErrCC implements MatchError {
+  // A list of Unicode codepoints / ranges of codepoints.
+  constructor(public charClasses: Array<number|[number, number]>) {}
+
+  public toGrammarString() {
+    return `[${
+        this.charClasses
+            .map((charClass) => {
+              return JSON
+                  .stringify(
+                      typeof charClass === 'number' ?
+                          String.fromCodePoint(charClass) :
+                          `${String.fromCodePoint(charClass[0])}-${
+                              String.fromCodePoint(charClass[1])}`)
+                  .slice(1, -1);
+            })
+            .join('')}]`;
+  }
+}
+
+// A failed wildcard match.
+export class ErrAny implements MatchError {
+  public toGrammarString() {
+    return '.';
+  }
+}
+
+export class ParseError {
+  constructor(
+      public pos: number, public line: number, public col: number,
+      public nt: string[], public chars: MatchError[]) {}
+
+  public toString(): string {
+    let chars;
+    chars = this.chars.map((err) => err.toGrammarString()).join(' | ');
```

Comment:
nit: is `err` intended to be e.g. `char` here?

Reply:
No, `chars` is an array of `MatchError`s. We should probably rename it in an upcoming PR, but not int this one as `ParseError` is a public class.